rs/registry/canister/src/mutations/node_management/do_replace_operator.rs

```diff
@@ -0,0 +1,829 @@
+                Ok(_) => panic!("Expected error, but got Ok."),
+                Err(e) => assert!(
+                    e.contains(expected),
+                    "Expected error containing '{expected}', but got '{e}'"
+                ),
+            }
+        }
+
+        #[track_caller]
+        fn assert_ok(self) {
+            assert!(
+                self.is_ok(),
+                "Expected Ok, but got Err: {}",
+                self.err().unwrap()
+            )
+        }
+    }
+
+    fn mutation(key: Vec<u8>, value: Vec<u8>) -> RegistryMutation {
+        RegistryMutation {
+            mutation_type: Type::Upsert as i32,
+            key,
+            value,
+        }
+    }
+
+    fn operator_mutation(
+        operator: PrincipalId,
+        provider: PrincipalId,
+        node_allowance: u64,
+        dc_id: &str,
+    ) -> RegistryMutation {
+        let operator_record = NodeOperatorRecord {
+            node_operator_principal_id: operator.as_slice().to_vec(),
+            node_allowance,
+            node_provider_principal_id: provider.as_slice().to_vec(),
+            dc_id: dc_id.to_string(),
+            ..Default::default()
+        };
+
+        mutation(
+            make_node_operator_record_key(operator).as_bytes().to_vec(),
+            operator_record.encode_to_vec(),
+        )
+    }
+
+    thread_local! {
+        /// Needed for invariant checks of each node
+        /// because each node has to have unique xnet
+        /// and http endpoints
+        static NODE_NUM: RefCell<u8> = const { RefCell::new(0) };
+    }
+
+    fn node_mutation(node_id: NodeId, operator: PrincipalId) -> RegistryMutation {
+        let current_node_num = NODE_NUM.with_borrow_mut(|current| {
+            *current += 1;
+            *current
+        });
+        let node_record = NodeRecord {
+            node_operator_id: operator.as_slice().to_vec(),
+            xnet: Some(ConnectionEndpoint {
+                ip_addr: format!("192.{current_node_num}.0.1"),
+                port: 8080,
+            }),
+            http: Some(ConnectionEndpoint {
+                ip_addr: format!("192.{current_node_num}.0.2"),
+                port: 8080,
+            }),
+            ..Default::default()
+        };
+
+        mutation(
+            make_node_record_key(node_id).as_bytes().to_vec(),
+            node_record.encode_to_vec(),
+        )
+    }
+
+    fn dc_mutation(dc_id: &str) -> RegistryMutation {
+        let dc_record = DataCenterRecord {
+            id: dc_id.to_string(),
+            ..Default::default()
+        };
+
+        mutation(
+            make_data_center_record_key(dc_id).as_bytes().to_vec(),
+            dc_record.encode_to_vec(),
+        )
+    }
+
+    #[test]
+    fn disallow_empty_node_ids() {
+        let mut registry = Registry::new();
+
+        registry
+            .do_replace_operator_(payload(operator(1), operator(2), &[]), caller(99))
+            .assert_err_contains("No nodes to update supplied");
```

Comment:
I like this test helper, it's pretty nifty.  I might steal this and put it in a library :)